### ISSUE TITLE
Batchnorm with CuDNN v5

### ISSUE THREAD

**Issue by fvisin** [Lasagne/Lasagne#743]
Hey guys,

@marcociccone and I would like to use the [new cuDNN 5 Theano batchnorm](https://github.com/Theano/Theano/blob/2711f775d34e453611c797c09cb3da1734f19c6e/theano/sandbox/cuda/dnn.py#L2659) with Lasagne. 

I read here and there some discussion about it (e.g. https://github.com/Lasagne/Lasagne/issues/575#issuecomment-228396513) but I coulnd't find any implementation.

I was wondering:

1. Is there any code for cuDNN 5 batchnorm anywhere?
2. Should we work on it, would you be interested in a PR?
3. My understanding is that you decided to keep the current BN interface. I didn't have much time to look into the theano Op nor your previous implementation of BN yet. Can you give us a quick idea of what/how you would like the new BN to be implemented?

Thanks!

**Comment by f0k** [comment]
> Is there any code for cuDNN 5 batchnorm anywhere?

Not in Lasagne, no.

> Should we work on it, would you be interested in a PR?

Sure, go ahead! We can only merge this into Lasagne 0.2 if there's a Theano version on PyPI that has my cudnn batchnorm Op in it, though, otherwise we'll have to release 0.2 first. We want release 0.2 to be compatible to a Theano version available on PyPI, so currently we're limited to features available in Theano 0.8.2.

> My understanding is that you decided to keep the current BN interface. I didn't have much time to look into the theano Op nor your previous implementation of BN yet. Can you give us a quick idea of what/how you would like the new BN to be implemented?

In its current form, the Theano training Op allows to compute the batch-normalized inputs, the input means and the input invstds at once. My idea would be to return the batch-normalized inputs as the layer output (obviously), and use the input means and invstds to update the exponential moving average statistics in the same way as done currently.
cuDNN would allow updating the exponential moving average statistics for us, but then it updates the means and variances, not means and inverse standard deviations, so we would have a different interface. Also it would require extending the Theano Op. So I'd leave that for later, if anything.

The Theano testing Op is not useful for our implementation, since it would require the variances, not the inverse standard deviations. I don't know if it would be faster than the vanilla Theano expression anyway.

For the concrete implementation, we need a new class BatchNormalizationDNNLayer in lasagne.layers.dnn that derives from BatchNormalizationLayer and overrides `get_output_for()`. It will also need to override `__init__()` to put some constraints on allowed combinations of input dimensionality and the `axes` argument, since cuDNN only supports up to 5D tensors, and only supports "per-activation" and "spatial" normalization. And then we need tests.

**Comment by fvisin** [comment]
Sounds great, thank you for the immediate reply!
We will try to find some spare time to work on it and open a PR as soon as we have something.

> Sure, go ahead! We can only merge this into Lasagne 0.2 if there's a Theano version on PyPI that has my cudnn batchnorm Op in it, though, otherwise we'll have to release 0.2 first. We want release 0.2 to be compatible to a Theano version available on PyPI, so currently we're limited to features available in Theano 0.8.2.

The plan with Theano was to try to have more frequent releases, I'll try to put some pressure on the next one. :)

**Comment by f0k** [comment]
> The Theano testing Op is not useful for our implementation, since it would require the variances, not the inverse standard deviations. I don't know if it would be faster than the vanilla Theano expression anyway.

Wrong, I know: https://github.com/Theano/Theano/pull/4582#issuecomment-224368422
Still, just having the training Op would already speed up things a lot.

> We will try to find some spare time to work on it and open a PR as soon as we have something.

Great. Just a gist with the new layer would be good for a start as well.

> The plan with Theano was to try to have more frequent releases

I can't really say anything given our release history...

**Comment by marcociccone** [comment]
I've started to implement the BatchNormDNNLayer: you can find the gist [here](https://gist.github.com/marcociccone/1f19a197df97871288535c4ed40927a0#file-batchnormdnnlayer-py-L63)

It's still WIP, in particular I would like your opinion on Line 63, I'm not sure what's the best way to check if we need to use the training op or the inference op.

Moreover, from the first tests, it doesn't that the cudnn operation dramatically improves the speed as I expected from your previous benchmark.. Could you please have a look at it?

**Comment by f0k** [comment]
> I've started to implement the BatchNormDNNLayer

Cool!

> It's still WIP, in particular I would like your opinion on Line 63, I'm not sure what's the best way to check if we need to use the training op or the inference op.

As stated above:
> The Theano testing Op is not useful for our implementation, since it would require the variances, not the inverse standard deviations.

So we'd never use the inference Op.

Your `if deterministic` currently checks whether we need the training Op at all. This should probably be `if update_averages or not use_averages`, since those are the cases where we'd either need the input statistics, or the input normalized by the input statistics. The `if update_averages` stays as before. In the end, you'll need to check `if use_averages` to decide whether to return the existing `normalized`, or compute it from `self.mean` and `self.inv_std`. The `deterministic` flag only gives defaults for `batch_norm_use_averages` and `batch_norm_update_averages` in case they're not given; it should not be checked for directly later on.

About `class BatchNormDNNLayer(Layer):`: You may want to derive it from the existing `BatchNormLayer`, so you don't need to replicate parameter initialization and such. Note that cuDNN does not allow to omit beta and gamma, though, if I remember correctly. Check whether the Op handles this, otherwise you'll have to create tensor constants to keep it happy.

> Moreover, from the first tests, it doesn't that the cudnn operation dramatically improves the speed as I expected from your previous benchmark.

Did you time the batch normalization separately or as part of a larger network?

**Comment by f0k** [comment]
> Moreover, from the first tests, it doesn't that the cudnn operation dramatically improves the speed as I expected from your previous benchmark.. Could you please have a look at it?

Okay, I took your code as a starting point, finished it and ran it. For a DenseNet (L=40, k=12) on a Tesla K40c, speed improves from 5:49 per epoch to 4:25. Yay!

My implementation is here: https://gist.github.com/f0k/8698d1a71215c9840b779048fff14e4b

I'd prefer to have `axes` as the second argument, for direct compatibility to the original batch normalization layer. You can infer the mode from the axes then, and raise an exception if there's no mode fitting the given combination of axes.

You're still very welcome to submit a PR along with tests, this would be highly appreciated!

**Comment by f0k** [comment]
The alternative would be adding the batch normalization training Op directly to Theano, and adding an optimization that replaces it with the cuDNN version if possible. This way we wouldn't need another Lasagne layer, and existing models would transparently use cuDNN. It's just that the interface (normalized, mean, inv_std) is a bit funny.

**Comment by marcociccone** [comment]
Thanks for your suggestions and for your implementation! Derive from the existent `BatchNormLayer` it's more elegant for sure.
I will finish the layer with the parameter `axes` and then I will submit the PR.

> Okay, I took your code as a starting point, finished it and ran it. For a DenseNet (L=40, k=12) on a Tesla K40c, speed improves from 5:49 per epoch to 4:25. On a GTX 970, it improves from 3:55 to 2:30. Yay!

Sweet! I will run a test on my network and I'll post the benchmark asap

**Comment by fvisin** [comment]
> The alternative would be adding the batch normalization training Op directly to Theano, and adding an optimization that replaces it with the cuDNN version if possible. This way we wouldn't need another Lasagne layer, and existing models would transparently use cuDNN. It's just that the interface (normalized, mean, inv_std) is a bit funny.

For the moment I think that proceeding with the PR is probably the best solution. A Theano Op + optimization would be ideal, but would take some time (and a volunteer :D) and I don't think there is a plan for it at the moment on Theano's side (@nouiz, could you comment on that?). On the other hand, from my understanding the code for BN with CuDNN 5 is ready so my suggestion is to do the PR and leave the Theano Op as a possible future improvement.

**Comment by f0k** [comment]
> On the other hand, from my understanding the code for BN with CuDNN 5 is ready so my suggestion is to do the PR and leave the Theano Op as a possible future improvement.

The code was almost ready, had to do some changes to make it work with beta=gamma=None. And rethinking this, we could add it even for Lasagne 0.2, but have it fail if the Theano version is too old. Travis' tests don't include cuDNN anyway, so we don't need to update the Theano dependency for this. I think I could do a PR myself later this week.
